scripts/size_report/report_binary_size.ts

```diff
@@ -60,109 +63,82 @@ function generateReportForCDNScripts(): Report[] {
   for (const file of files) {
     const { size } = fs.statSync(file);
     const fileName = file.split('/').slice(-1)[0];
-    reports.push(makeReportObject('firebase', fileName, size));
+    reports.push({ sdk: 'firebase', type: fileName, value: size });
   }
 
   return reports;
 }
 
-// NPM packages
 async function generateReportForNPMPackages(): Promise<Report[]> {
   const reports: Report[] = [];
-  const fields = [
-    'main',
-    'module',
-    'esm2017',
-    'browser',
-    'react-native',
-    'lite',
-    'lite-esm2017'
-  ];
-
   const packageInfo = JSON.parse(
     execSync('npx lerna ls --json --scope @firebase/*').toString()
   );
-
-  const taskPromises: Promise<void>[] = [];
-  for (const pkg of packageInfo) {
-    // we traverse the dir in order to include binaries for submodules, e.g. @firebase/firestore/memory
-    // Currently we only traverse 1 level deep because we don't have any submodule deeper than that.
-    traverseDirs(pkg.location, collectBinarySize, 0, 1);
+  for (const info of packageInfo) {
+    const packages = await findAllPackages(info.location);
+    for (const pkg of packages) {
+      reports.push(...(await collectBinarySize(pkg)));
+    }
   }
-
-  await Promise.all(taskPromises);
-
   return reports;
+}
 
-  function collectBinarySize(path: string) {
-    const packageJsonPath = `${path}/package.json`;
-    if (!fs.existsSync(packageJsonPath)) {
-      return;
-    }
-
-    const promise = new Promise<void>(async resolve => {
-      const packageJson = require(packageJsonPath);
-
-      for (const field of fields) {
-        if (packageJson[field]) {
-          const filePath = pathResolve(path, packageJson[field]);
-          // Need to create a bundle and get the size of the bundle instead of reading the size of the file directly.
-          // It is because some packages might be split into multiple files in order to share code between entry points.
-          const bundle = await rollup.rollup({
-            input: filePath,
-            plugins: [commonjs()]
-          });
-
-          const { output } = await bundle.generate({ format: 'es' });
-          const rawCode = output[0].code;
-
-          // remove comments and whitespaces, then get size
-          const { code } = await terser.minify(rawCode, {
-            format: {
-              comments: false
-            },
-            mangle: false,
-            compress: false
-          });
-
-          const size = Buffer.byteLength(code!, 'utf-8');
-          reports.push(makeReportObject(packageJson.name, field, size));
+async function findAllPackages(root: string): Promise<string[]> {
+  return new Promise((resolve, reject) => {
+    glob(
+      '**/package.json',
```

Comment:
Can we exclude package.json that belong to demo projects, e.g. `packages-exp/auth-exp/demo`? I have no idea how it worked before, because it seems it will break the script.

Reply:
`packages-exp/auth-exp` is not included in the below list, and is actually not involved here. But I definitely can add more paths here that we are sure should be excluded from the size report. Any other paths that you think should be excluded as well?

```
yifany:firebase-js-sdk yifany$ npx lerna ls --json --scope @firebase/*
lerna notice cli v3.22.1
lerna info versioning independent
lerna notice filter including "@firebase/*"
lerna info filter [ '@firebase/*' ]
[
  {
    "name": "@firebase/analytics-exp",
    "version": "0.0.900",
    "private": false,
    "location": "/Users/yifany/Documents/firebase/firebase-js-sdk/packages-exp/analytics-exp"
  },
  {
    "name": "@firebase/analytics-interop-types",
    "version": "0.1.5",
    "private": false,
    "location": "/Users/yifany/Documents/firebase/firebase-js-sdk/packages/analytics-interop-types"
  },
  {
    "name": "@firebase/analytics-types",
    "version": "0.4.0",
    "private": false,
    "location": "/Users/yifany/Documents/firebase/firebase-js-sdk/packages/analytics-types"
  },
  {
    "name": "@firebase/analytics",
    "version": "0.6.2",
    "private": false,
    "location": "/Users/yifany/Documents/firebase/firebase-js-sdk/packages/analytics"
  },
  {
    "name": "@firebase/app-types",
    "version": "0.6.1",
    "private": false,
    "location": "/Users/yifany/Documents/firebase/firebase-js-sdk/packages/app-types"
  },
  {
    "name": "@firebase/app",
    "version": "0.6.14",
    "private": false,
    "location": "/Users/yifany/Documents/firebase/firebase-js-sdk/packages/app"
  },
  {
    "name": "@firebase/auth-interop-types",
    "version": "0.1.5",
    "private": false,
    "location": "/Users/yifany/Documents/firebase/firebase-js-sdk/packages/auth-interop-types"
  },
  {
    "name": "@firebase/auth-types",
    "version": "0.10.1",
    "private": false,
    "location": "/Users/yifany/Documents/firebase/firebase-js-sdk/packages/auth-types"
  },
  {
    "name": "@firebase/auth",
    "version": "0.16.2",
    "private": false,
    "location": "/Users/yifany/Documents/firebase/firebase-js-sdk/packages/auth"
  },
  {
    "name": "@firebase/component",
    "version": "0.1.21",
    "private": false,
    "location": "/Users/yifany/Documents/firebase/firebase-js-sdk/packages/component"
  },
  {
    "name": "@firebase/database-types",
    "version": "0.6.1",
    "private": false,
    "location": "/Users/yifany/Documents/firebase/firebase-js-sdk/packages/database-types"
  },
  {
    "name": "@firebase/database",
    "version": "0.9.1",
    "private": false,
    "location": "/Users/yifany/Documents/firebase/firebase-js-sdk/packages/database"
  },
  {
    "name": "@firebase/firestore-types",
    "version": "2.1.0",
    "private": false,
    "location": "/Users/yifany/Documents/firebase/firebase-js-sdk/packages/firestore-types"
  },
  {
    "name": "@firebase/firestore",
    "version": "2.1.4",
    "private": false,
    "location": "/Users/yifany/Documents/firebase/firebase-js-sdk/packages/firestore"
  },
  {
    "name": "@firebase/functions-types",
    "version": "0.4.0",
    "private": false,
    "location": "/Users/yifany/Documents/firebase/firebase-js-sdk/packages/functions-types"
  },
  {
    "name": "@firebase/functions",
    "version": "0.6.1",
    "private": false,
    "location": "/Users/yifany/Documents/firebase/firebase-js-sdk/packages/functions"
  },
  {
    "name": "@firebase/installations-types",
    "version": "0.3.4",
    "private": false,
    "location": "/Users/yifany/Documents/firebase/firebase-js-sdk/packages/installations-types"
  },
  {
    "name": "@firebase/installations",
    "version": "0.4.19",
    "private": false,
    "location": "/Users/yifany/Documents/firebase/firebase-js-sdk/packages/installations"
  },
  {
    "name": "@firebase/logger",
    "version": "0.2.6",
    "private": false,
    "location": "/Users/yifany/Documents/firebase/firebase-js-sdk/packages/logger"
  },
  {
    "name": "@firebase/messaging-types",
    "version": "0.5.0",
    "private": false,
    "location": "/Users/yifany/Documents/firebase/firebase-js-sdk/packages/messaging-types"
  },
  {
    "name": "@firebase/messaging",
    "version": "0.7.3",
    "private": false,
    "location": "/Users/yifany/Documents/firebase/firebase-js-sdk/packages/messaging"
  },
  {
    "name": "@firebase/performance-types",
    "version": "0.0.13",
    "private": false,
    "location": "/Users/yifany/Documents/firebase/firebase-js-sdk/packages/performance-types"
  },
  {
    "name": "@firebase/performance",
    "version": "0.4.5",
    "private": false,
    "location": "/Users/yifany/Documents/firebase/firebase-js-sdk/packages/performance"
  },
  {
    "name": "@firebase/polyfill",
    "version": "0.3.36",
    "private": false,
    "location": "/Users/yifany/Documents/firebase/firebase-js-sdk/packages/polyfill"
  },
  {
    "name": "@firebase/remote-config-types",
    "version": "0.1.9",
    "private": false,
    "location": "/Users/yifany/Documents/firebase/firebase-js-sdk/packages/remote-config-types"
  },
  {
    "name": "@firebase/remote-config",
    "version": "0.1.30",
    "private": false,
    "location": "/Users/yifany/Documents/firebase/firebase-js-sdk/packages/remote-config"
  },
  {
    "name": "@firebase/rules-unit-testing",
    "version": "1.1.10",
    "private": false,
    "location": "/Users/yifany/Documents/firebase/firebase-js-sdk/packages/rules-unit-testing"
  },
  {
    "name": "@firebase/storage-types",
    "version": "0.3.13",
    "private": false,
    "location": "/Users/yifany/Documents/firebase/firebase-js-sdk/packages/storage-types"
  },
  {
    "name": "@firebase/storage",
    "version": "0.4.2",
    "private": false,
    "location": "/Users/yifany/Documents/firebase/firebase-js-sdk/packages/storage"
  },
  {
    "name": "@firebase/util",
    "version": "0.3.4",
    "private": false,
    "location": "/Users/yifany/Documents/firebase/firebase-js-sdk/packages/util"
  },
  {
    "name": "@firebase/webchannel-wrapper",
    "version": "0.4.1",
    "private": false,
    "location": "/Users/yifany/Documents/firebase/firebase-js-sdk/packages/webchannel-wrapper"
  }
]
lerna success found 31 packages
```

Reply:
Ah it excludes all private packages, so we should be fine. The path is going to change when exp becomes official, so let's hold off adding the path until then.